File: modules/extensions/underline_passthrough.rb
# underline_passthrough.rb
require 'asciidoctor'
require 'asciidoctor/extensions'

Asciidoctor::Extensions.register do
  preprocessor do
    process do |document, reader|
      # Transform every span. Adjust the regex if your spans span multiple lines.
      new_lines = reader.lines.map do |line|
        line.gsub(/<span class="underline">(.+?)<\/span>/) do
          content = Regexp.last_match(1)
          # Wrap in triple-plus passthrough to inhibit all substitutions
          "pass:[+++<span class=\"underline\">#{content}</span>+++]"
        end
      end
      Asciidoctor::Reader.new(new_lines, document)
    end
  end
end

File: modules/extensions/underline_passthrough.js
module.exports = function registerUnderlinePassthrough(registry) {
  registry.preprocessor('underline-passthrough', function () {
    this.process((doc, reader) => {
      // Read all lines from the reader
      const lines = [];
      let line;
      while ((line = reader.readLine()) !== undefined) {
        // Transform lines containing span.underline
        const transformedLine = line.replace(/<span class="underline">(.+?)<\/span>/g, (match, content) => {
          return `pass:[+++<span class="underline">${content}</span>+++]`;
        });
        lines.push(transformedLine);
      }

      // Replace the reader's content
      reader.pushInclude(
        lines.join('\n'),
        'virtual.adoc',
        1
      );
      return reader;
    });
  });
}; 